File: internal/httpapi/silence.go
package httpapi

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"time"

	"github.com/cropalato/promview/internal/alertmanager"
	"github.com/cropalato/promview/internal/alerts"
	"github.com/cropalato/promview/internal/auth"
)

/*
Silencing is the only action here that hides alerts rather than surfacing them,
and the only one that writes to a system promview does not own. Three things
follow from that, and they are enforced here rather than left to the caller:

  - It needs operator rights, and the scope query re-checks them in SQL, so an
    operator scoped to one team cannot silence another team's alerts by naming a
    group that spans both.
  - It is always attributed to the signed-in person. Open mode has no user and
    no operator rights, so it cannot reach this at all.
  - It always ends. The window is bounded by configuration, because a silence
    long enough to outlast the on-call rotation is a deleted rule wearing a
    disguise.
*/

const maxSilenceBodyBytes = 4096

// Silencer creates a silence on one Alertmanager. Narrow on purpose: the
// transport should not be able to read alerts back out of the Alertmanager.
type Silencer interface {
	CreateSilence(ctx context.Context, baseURL string, token string, silence alertmanager.Silence) (string, error)
}

type silenceRequest struct {
	// Seconds the silence lasts. Zero or absent takes the deployment default.
	DurationSeconds int64  `json:"durationSeconds"`
	Comment         string `json:"comment"`
}

type groupSilenceRequest struct {
	GroupBy []string          `json:"groupBy"`
	Key     map[string]string `json:"key"`
	// ExpectedMatchers is the match the console showed before confirming. The
	// resolved match is narrower than the grouping key and depends on which
	// members are firing right now, so a member joining between the preview and
	// the confirm would silence more than the operator read. Sending back what
	// was read turns that race into a 409 instead of a surprise.
	ExpectedMatchers map[string]string `json:"expectedMatchers"`
	DurationSeconds  int64             `json:"durationSeconds"`
	Comment          string            `json:"comment"`
}

type groupSilencePreviewRequest struct {
	GroupBy []string          `json:"groupBy"`
	Key     map[string]string `json:"key"`
}

func (api *API) silenceAlert(w http.ResponseWriter, r *http.Request) {
	principal, ok := api.silenceRequestPrincipal(w, r)
	if !ok {
		return
	}
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil || id < 1 {
		writeError(w, http.StatusBadRequest, "alert id is invalid")
		return
	}
	var body silenceRequest
	if !decodeSilenceBody(w, r, &body) {
		return
	}
	scope, err := api.store.SilenceScopeForAlert(r.Context(), principal, id)
	if !api.writeScopeError(w, err) {
		return
	}
	api.createSilences(w, r, principal, scope, body.DurationSeconds, body.Comment)
}

func (api *API) silenceGroup(w http.ResponseWriter, r *http.Request) {
	principal, ok := api.silenceRequestPrincipal(w, r)
	if !ok {
		return
	}
	var body groupSilenceRequest
	if !decodeSilenceBody(w, r, &body) {
		return
	}
	if len(body.GroupBy) == 0 || len(body.Key) == 0 {
		writeError(w, http.StatusBadRequest, "groupBy and key are required")
		return
	}
	scope, err := api.store.SilenceScopeForGroup(r.Context(), principal, body.GroupBy, body.Key)
	if !api.writeScopeError(w, err) {
		return
	}
	if body.ExpectedMatchers != nil && !sameLabels(body.ExpectedMatchers, scope.Labels) {
		writeJSON(w, http.StatusConflict, map[string]any{
			"error":    "the group changed since it was previewed; review the new match and confirm again",
			"matchers": scope.Labels,
			"targets":  silenceTargetsResponse(scope.Targets),
		})
		return
	}
	api.createSilences(w, r, principal, scope, body.DurationSeconds, body.Comment)
}

// previewGroupSilence answers what a group silence would actually match. The
// console cannot work this out for itself: the match is the labels the firing
// members agree on, which only the store knows, and confirming a dialog that
// showed the grouping key instead would understate what is about to be hidden.
func (api *API) previewGroupSilence(w http.ResponseWriter, r *http.Request) {
	principal, ok := api.silenceRequestPrincipal(w, r)
	if !ok {
		return
	}
	var body groupSilencePreviewRequest
	if !decodeSilenceBody(w, r, &body) {
		return
	}
	if len(body.GroupBy) == 0 || len(body.Key) == 0 {
		writeError(w, http.StatusBadRequest, "groupBy and key are required")
		return
	}
	scope, err := api.store.SilenceScopeForGroup(r.Context(), principal, body.GroupBy, body.Key)
	if !api.writeScopeError(w, err) {
		return
	}
	members := 0
	for _, target := range scope.Targets {
		members += target.Members
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"matchers":    scope.Labels,
		"memberCount": members,
		"targets":     silenceTargetsResponse(scope.Targets),
	})
}

func silenceTargetsResponse(targets []alerts.SilenceTarget) []map[string]any {
	response := make([]map[string]any, 0, len(targets))
	for _, target := range targets {
		response = append(response, map[string]any{
			"source":   target.Source,
			"matchers": target.Labels,
			"members":  target.Members,
		})
	}
	return response
}

func sameLabels(left, right map[string]string) bool {
	if len(left) != len(right) {
		return false
	}
	for name, value := range left {
		if other, ok := right[name]; !ok || other != value {
			return false
		}
	}
	return true
}

// silenceRequestPrincipal runs the checks every silence shares: who is asking,
// whether they may operate, and whether the request came from this console.
func (api *API) silenceRequestPrincipal(w http.ResponseWriter, r *http.Request) (auth.Principal, bool) {
	principal, ok := requestPrincipal(r)
	if !ok {
		writeError(w, http.StatusInternalServerError, "principal is unavailable")
		return auth.Principal{}, false
	}
	if api.silencer == nil {
		writeError(w, http.StatusNotImplemented, "silencing is not configured")
		return auth.Principal{}, false
	}
	if !principal.CanOperate() {
		writeError(w, http.StatusForbidden, "operator access required")
		return auth.Principal{}, false
	}
	if !validMutationOrigin(r) {
		writeError(w, http.StatusForbidden, "invalid request origin")
		return auth.Principal{}, false
	}
	if silenceAuthor(principal) == "" {
		// Every silence is attributable or it is not created. Nothing should
		// reach here — operator rights imply a user — but an unattributed
		// silence is exactly the artefact nobody can later explain.
		writeError(w, http.StatusForbidden, "a silence needs a named author")
		return auth.Principal{}, false
	}
	return principal, true
}

func decodeSilenceBody(w http.ResponseWriter, r *http.Request, into any) bool {
	decoder := json.NewDecoder(http.MaxBytesReader(w, r.Body, maxSilenceBodyBytes))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(into); err != nil && !errors.Is(err, io.EOF) {
		writeError(w, http.StatusBadRequest, "request body is invalid")
		return false
	}
	if decoder.Decode(&struct{}{}) != io.EOF {
		writeError(w, http.StatusBadRequest, "request body is invalid")
		return false
	}
	return true
}

// writeScopeError maps a scope failure onto a response, returning false when it
// wrote one.
func (api *API) writeScopeError(w http.ResponseWriter, err error) bool {
	switch {
	case err == nil:
		return true
	case errors.Is(err, alerts.ErrNotFound):
		writeError(w, http.StatusNotFound, "alert not found")
	case errors.Is(err, alerts.ErrNoSilenceTarget):
		writeError(w, http.StatusConflict, "no source in scope has an alertmanager url configured")
	default:
		writeError(w, http.StatusBadRequest, err.Error())
	}
	return false
}

func (api *API) createSilences(
	w http.ResponseWriter,
	r *http.Request,
	principal auth.Principal,
	scope alerts.SilenceScope,
	durationSeconds int64,
	comment string,
) {
	window, err := api.silenceWindow(durationSeconds)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	startsAt := time.Now().UTC()
	endsAt := startsAt.Add(window)
	author := silenceAuthor(principal)

	results := make([]alerts.SilenceResult, 0, len(scope.Targets))
	created := 0
	for _, target := range scope.Targets {
		// One silence body per target: each Alertmanager gets the narrowest
		// match covering its own members, which is usually narrower than what
		// the targets have in common. See alerts.SilenceTarget.
		silence := alertmanager.Silence{
			Matchers:  alertmanager.MatchersFromLabels(target.Labels),
			StartsAt:  startsAt,
			EndsAt:    endsAt,
			CreatedBy: author,
			Comment:   comment,
		}
		id, err := api.silencer.CreateSilence(r.Context(), target.AlertmanagerURL, target.AlertmanagerToken, silence)
		if err != nil {
			// One Alertmanager refusing must not hide that the others accepted;
			// an operator who thinks a group is handled when half of it is still
			// firing is worse off than one told exactly which half failed.
			results = append(results, alerts.SilenceResult{
				Source:   target.Source,
				Matchers: target.Labels,
				Members:  target.Members,
				Error:    err.Error(),
			})
			continue
		}
		created++
		results = append(results, alerts.SilenceResult{
			Source:    target.Source,
			SilenceID: id,
			Matchers:  target.Labels,
			Members:   target.Members,
		})
		// Alertmanager expires the silence and forgets why it existed. Recording
		// it here is what still answers "who silenced this" afterwards, and a
		// failure to record must not turn a silence that worked into an error.
		if err := api.store.RecordSilence(r.Context(), alerts.SilenceRecord{
			Source:    target.Source,
			SilenceID: id,
			Matchers:  target.Labels,
			CreatedBy: author,
			Comment:   comment,
			StartsAt:  startsAt,
			EndsAt:    endsAt,
		}); err != nil {
			slog.Error("could not record a created silence",
				"source", target.Source, "silence", id, "error", err)
		}
	}

	status := http.StatusCreated
	switch {
	case created == 0:
		status = http.StatusBadGateway
	case created < len(scope.Targets):
		status = http.StatusMultiStatus
	}
	writeJSON(w, status, map[string]any{
		"endsAt":    endsAt.Format(time.RFC3339),
		"createdBy": author,
		"matchers":  alertmanager.MatchersFromLabels(scope.Labels),
		"results":   results,
	})
}

// silenceWindow resolves the requested duration against the deployment's
// default and ceiling.
func (api *API) silenceWindow(durationSeconds int64) (time.Duration, error) {
	if durationSeconds == 0 {
		return api.config.SilenceDefaultDuration, nil
	}
	if durationSeconds < 0 {
		return 0, errors.New("durationSeconds must be positive")
	}
	window := time.Duration(durationSeconds) * time.Second
	if window > api.config.SilenceMaxDuration {
		return 0, errors.New("durationSeconds exceeds the configured maximum silence duration")
	}
	return window, nil
}

// silenceAuthor is who the Alertmanager will record. Email first: it is what
// identifies a person across the tools an incident actually spans.
func silenceAuthor(principal auth.Principal) string {
	if principal.Anonymous {
		return ""
	}
	for _, candidate := range []string{principal.Email, principal.Subject, principal.DisplayName} {
		if candidate != "" {
			return candidate
		}
	}
	return ""
}

File: web/src/alerts/silence.ts
/**
 * Silencing an alert on its Alertmanager.
 *
 * This is the only action in the console that hides alerts rather than
 * surfacing them, and the only one that writes to a system promview does not
 * own. The server decides who may do it, who it is attributed to, and how long
 * it may last; the console's job is to say clearly what is about to be silenced
 * and for how long, and to report honestly when only part of it worked.
 */

import { apiUrl } from '../config/apiBase';
import { apiFetch } from '../config/transport';

export const ALERT_SILENCE_URL = (id: string) => `/api/v1/alerts/${encodeURIComponent(id)}/silence`;
export const GROUP_SILENCE_URL = '/api/v1/groups/silence';
export const GROUP_SILENCE_PREVIEW_URL = '/api/v1/groups/silence/preview';

/** What happened at one Alertmanager. A group can span several. */
export interface SilenceResult {
  source: string;
  silenceId?: string;
  /** The exact match written here, which can differ between Alertmanagers. */
  matchers: Record<string, string>;
  members: number;
  error?: string;
}

export interface SilenceResponse {
  endsAt: string;
  createdBy: string;
  results: SilenceResult[];
}

/**
 * What a group silence would actually match, resolved by the server.
 *
 * The console cannot work this out: the match is every label the group's
 * firing members agree on, which is narrower than the grouping key and is only
 * knowable from the members themselves. Showing the key instead would
 * understate what is about to be hidden — a group keyed on `alertname` alone
 * would read as "silence this alert here" while silencing it everywhere.
 */
export interface SilenceTargetPreview {
  source: string;
  matchers: Record<string, string>;
  members: number;
}

export interface SilencePreview {
  /** What every target agrees on; each target's own match is at least this narrow. */
  matchers: Record<string, string>;
  memberCount: number;
  targets: SilenceTargetPreview[];
}

export interface SilenceRequest {
  durationSeconds: number;
  comment: string;
}

export class SilenceError extends Error {
  readonly status: number;
  /**
   * The match the server resolved, when it refused because the group moved
   * under the operator. Carrying it back lets the dialog show the new scope
   * instead of asking them to start again blind.
   */
  readonly matchers?: Record<string, string>;

  constructor(message: string, status: number, matchers?: Record<string, string>) {
    super(message);
    this.name = 'SilenceError';
    this.status = status;
    this.matchers = matchers;
  }
}

/** True when the group changed between the preview and the confirmation. */
export function isSilenceConflict(error: unknown): boolean {
  return error instanceof SilenceError && error.status === 409;
}

/** Duration choices offered in the dialog, in seconds. */
export const SILENCE_DURATIONS: readonly { seconds: number; label: string }[] = [
  { seconds: 30 * 60, label: '30 minutes' },
  { seconds: 60 * 60, label: '1 hour' },
  { seconds: 2 * 60 * 60, label: '2 hours' },
  { seconds: 4 * 60 * 60, label: '4 hours' },
  { seconds: 12 * 60 * 60, label: '12 hours' },
  { seconds: 24 * 60 * 60, label: '1 day' },
  { seconds: 7 * 24 * 60 * 60, label: '1 week' },
];

/**
 * The choices this deployment allows, with its own default guaranteed present.
 * A configured default of 90 minutes is not in the fixed list, and offering a
 * dialog that cannot express the deployment's own default would be absurd.
 */
export function silenceDurationOptions(
  defaultSeconds: number,
  maxSeconds: number,
): { seconds: number; label: string }[] {
  const options = SILENCE_DURATIONS.filter((option) => option.seconds <= maxSeconds);
  if (!options.some((option) => option.seconds === defaultSeconds)) {
    options.push({ seconds: defaultSeconds, label: formatDuration(defaultSeconds) });
  }
  return options.sort((left, right) => left.seconds - right.seconds);
}

export function formatDuration(seconds: number): string {
  if (seconds % (24 * 60 * 60) === 0) {
    const days = seconds / (24 * 60 * 60);
    return days === 1 ? '1 day' : `${days} days`;
  }
  if (seconds % (60 * 60) === 0) {
    const hours = seconds / (60 * 60);
    return hours === 1 ? '1 hour' : `${hours} hours`;
  }
  const minutes = Math.round(seconds / 60);
  return minutes === 1 ? '1 minute' : `${minutes} minutes`;
}

/**
 * The transport seam the desktop client needs: a Tauri shell owns credentials
 * in the Rust core and keeps them out of the webview, so it supplies its own
 * caller rather than inheriting the browser's cookie jar. Defaults to a
 * same-origin browser fetch, which is what the console itself uses.
 */
type FetchLike = (url: string, init?: RequestInit) => Promise<Response>;

async function postSilence(
  url: string,
  body: unknown,
  fetchImpl: FetchLike,
): Promise<SilenceResponse> {
  let response: Response;
  try {
    response = await fetchImpl(apiUrl(url), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
  } catch {
    throw new SilenceError('Unable to reach the Promview API', 0);
  }
  let payload: unknown = null;
  try {
    payload = await response.json();
  } catch {
    payload = null;
  }
  // 207 is a partial success: some Alertmanagers took the silence and some did
  // not. It is not an error, and the per-target results carry which is which.
  if (!response.ok && response.status !== 207) {
    const message =
      typeof payload === 'object' &&
      payload !== null &&
      typeof (payload as { error?: unknown }).error === 'string'
        ? (payload as { error: string }).error
        : `Silence failed (HTTP ${response.status})`;
    throw new SilenceError(message, response.status, parseMatchers(payload));
  }
  return parseSilenceResponse(payload);
}

export function parseSilenceResponse(payload: unknown): SilenceResponse {
  const raw = (typeof payload === 'object' && payload !== null ? payload : {}) as Record<
    string,
    unknown
  >;
  const results = Array.isArray(raw.results)
    ? raw.results.flatMap((entry): SilenceResult[] => {
        if (typeof entry !== 'object' || entry === null) {
          return [];
        }
        const item = entry as Record<string, unknown>;
        return [
          {
            source: typeof item.source === 'string' ? item.source : 'unknown',
            silenceId: typeof item.silenceId === 'string' ? item.silenceId : undefined,
            matchers: labelRecord(item.matchers),
            members:
              typeof item.members === 'number' && Number.isFinite(item.members) ? item.members : 0,
            error: typeof item.error === 'string' ? item.error : undefined,
          },
        ];
      })
    : [];
  return {
    endsAt: typeof raw.endsAt === 'string' ? raw.endsAt : '',
    createdBy: typeof raw.createdBy === 'string' ? raw.createdBy : '',
    results,
  };
}

export function silenceAlert(
  id: string,
  request: SilenceRequest,
  fetchImpl: FetchLike = apiFetch,
): Promise<SilenceResponse> {
  return postSilence(ALERT_SILENCE_URL(id), request, fetchImpl);
}

export function silenceGroup(
  groupBy: readonly string[],
  key: Record<string, string>,
  request: SilenceRequest,
  expectedMatchers?: Record<string, string>,
  fetchImpl: FetchLike = apiFetch,
): Promise<SilenceResponse> {
  return postSilence(
    GROUP_SILENCE_URL,
    { groupBy: [...groupBy], key, expectedMatchers, ...request },
    fetchImpl,
  );
}

/** Asks the server what a group silence would match, before offering to write it. */
export async function previewGroupSilence(
  groupBy: readonly string[],
  key: Record<string, string>,
  fetchImpl: FetchLike = apiFetch,
): Promise<SilencePreview> {
  let response: Response;
  try {
    response = await fetchImpl(apiUrl(GROUP_SILENCE_PREVIEW_URL), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ groupBy: [...groupBy], key }),
    });
  } catch {
    throw new SilenceError('Unable to reach the Promview API', 0);
  }
  let payload: unknown = null;
  try {
    payload = await response.json();
  } catch {
    payload = null;
  }
  if (!response.ok) {
    const message =
      typeof payload === 'object' &&
      payload !== null &&
      typeof (payload as { error?: unknown }).error === 'string'
        ? (payload as { error: string }).error
        : `Silence preview failed (HTTP ${response.status})`;
    throw new SilenceError(message, response.status);
  }
  return parseSilencePreview(payload);
}

export function parseSilencePreview(payload: unknown): SilencePreview {
  const raw = (typeof payload === 'object' && payload !== null ? payload : {}) as Record<
    string,
    unknown
  >;
  const targets = Array.isArray(raw.targets)
    ? raw.targets.flatMap((entry): SilenceTargetPreview[] => {
        if (typeof entry !== 'object' || entry === null) {
          return [];
        }
        const item = entry as Record<string, unknown>;
        return [
          {
            source: typeof item.source === 'string' ? item.source : 'unknown',
            matchers: labelRecord(item.matchers),
            members:
              typeof item.members === 'number' && Number.isFinite(item.members) ? item.members : 0,
          },
        ];
      })
    : [];
  return {
    matchers: labelRecord(raw.matchers),
    memberCount:
      typeof raw.memberCount === 'number' && Number.isFinite(raw.memberCount) ? raw.memberCount : 0,
    targets,
  };
}

function labelRecord(value: unknown): Record<string, string> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return {};
  }
  const result: Record<string, string> = {};
  for (const [name, entry] of Object.entries(value as Record<string, unknown>)) {
    if (typeof entry === 'string') {
      result[name] = entry;
    }
  }
  return result;
}

function parseMatchers(payload: unknown): Record<string, string> | undefined {
  if (typeof payload !== 'object' || payload === null) {
    return undefined;
  }
  const matchers = (payload as Record<string, unknown>).matchers;
  if (typeof matchers !== 'object' || matchers === null || Array.isArray(matchers)) {
    return undefined;
  }
  return labelRecord(matchers);
}
